Arrowgene.Ddon.GameServer/Handler/ConnectionLoginHandler.cs

```diff
@@ -109,7 +109,7 @@ public override void Handle(GameClient client, StructurePacket<C2SConnectionLogi
             Logger.Info(client, "Logged Into GameServer");
 
             // update login token for client
-            client.Account.LoginToken = GameToken.GenerateLoginToken();
```

Comment:
This one might be a bit tricky, and am not sure if we found the right solution yet, hence the issues.

As far as I am aware the client  deals with two kinds of tokens:
`OneTimeToken` - this is used for login authentication.
`SessionKey` - used to communicate state between login and game server.

These are used here:

OneTimeToken:
Login | ClientLoginHandler
Game | ConnectionLoginHandler

SessionKey:
Login | ConnectionLoginHandler
Login | ClientGetGameSessionKeyHandler
GameServer | ConnectionMoveInServerHandler
GameServer | ConnectionMoveOutServerHandler


Here is the order of packets during a normal login:
```
022-09-22 10:04:04 - Info - Consumer`1: Connected: [LoginClient@127.0.0.1:61295]
Client #000000 (1.0.1) C2L_CLIENT_CHALLENGE_REQ
Server #000000 (1.0.2) L2C_CLIENT_CHALLENGE_RES
Client #000001 (0.1.1) C2L_LOGIN_REQ   <---------------ClientLoginHandler
Server #000000 (0.1.2) L2C_LOGIN_RES   <---------------ClientLoginHandler
Client #000003 (3.0.1) C2L_GET_ERROR_MESSAGE_LIST_REQ
Server #000001 (3.0.16) L2C_GET_ERROR_MESSAGE_LIST_NTC
Server #000000 (3.0.2) L2C_GET_ERROR_MESSAGE_LIST_RES
Client #000004 (2.2.1) C2L_GET_LOGIN_SETTING_REQ
Server #000000 (2.2.2) L2C_GET_LOGIN_SETTING_RES
Client #000005 (4.0.1) C2L_GP_COURSE_GET_INFO_REQ
Server #000000 (4.0.2) L2C_GP_COURSE_GET_INFO_RES->Dump_22
Client #000006 (5.0.1) C2L_GET_CHARACTER_LIST_REQ
Server #000000 (5.0.2) L2C_GET_CHARACTER_LIST_RES
Client #000010 (5.3.1) C2L_CREATE_CHARACTER_DATA_REQ
Server #000002 (5.3.16) L2C_CREATE_CHARACTER_DATA_NTC
Server #000000 (5.3.2) L2C_CREATE_CHARACTER_DATA_RES
Client #000011 (5.1.1) C2L_DECIDE_CHARACTER_ID_REQ
Server #000000 (5.1.2) L2C_DECIDE_CHARACTER_ID_RES
Server #000003 (5.5.16) L2C_LOGIN_WAIT_NUM_NTC
Server #000004 (2.3.16) L2C_NEXT_CONNECT_SERVER_NTC
Client #000012 (2.1.1) C2L_GET_GAME_SESSION_KEY_REQ
Server #000000 (2.1.2) L2C_GET_GAME_SESSION_KEY_RES
2022-09-22 10:04:39 - Info - Consumer`1: Connected: [GameClient@127.0.0.1:61296]
Client #000000 (44.0.1) C2S_CERT_CLIENT_CHALLENGE_REQ
Server #000000 (44.0.2) S2C_CERT_CLIENT_CHALLENGE_RES
Client #000001 (0.1.1) C2S_CONNECTION_LOGIN_REQ <--------------------------ConnectionLoginHandler
Server #000000 (0.1.2) S2C_CONNECTION_LOGIN_RES <--------------------------ConnectionLoginHandler
Client #000003 (1.4.1) C2S_SERVER_GET_REAL_TIME_REQ
Server #000000 (1.4.2) S2C_SERVER_GET_REAL_TIME_RES
Client #000013 (0.2.1) C2L_LOGOUT_REQ
Server #000000 (0.2.2) L2C_LOGOUT_RES
2022-09-22 10:04:40 - Info - Consumer`1: Disconnected: [LoginClient@127.0.0.1:61295][Acc:aaaaaaaa00000000ffee]
Client #000004 (1.1.1) C2S_SERVER_GET_GAME_SETTING_REQ
Server #000000 (1.1.2) S2C_SERVER_GET_GAME_SETTING_RES->Dump_10
Client #000005 (2.0.1) C2S_CHARACTER_DECIDE_CHARACTER_ID_REQ
Server #000000 (2.0.2) S2C_CHARACTER_DECIDE_CHARACTER_ID_RES
Server #000001 (2.6.16) S2C_CHARACTER_CONTENTS_RELEASE_ELEMENT_NTC
Client #000006 (31.17.1) C2S_CLAN_CLAN_SETTING_UPDATE_REQ
Server #000000 (31.17.2) S2C_CLAN_CLAN_SETTING_UPDATE_RES->Dump_18
Client #000007 (12.0.1) C2S_STAGE_GET_STAGE_LIST_REQ
Server #000000 (12.0.2) S2C_STAGE_GET_STAGE_LIST_RES
[...] pressed close button
Client #000030 (0.2.1) C2S_CONNECTION_LOGOUT_REQ
Server #000000 (0.2.2) S2C_CONNECTION_LOGOUT_RES
2022-09-22 10:04:44 - Info - Consumer`1: Disconnected: [GameClient@127.0.0.1:61296][Acc:(1)aaaaaaaa00000000ffee][Cha:(1)Sad Sdsd]
```

I assume the `OneTimeToken` is updated by the game server for security reasons, or ensure that it can not be used to login again. From my observation is that if you come back to title screen, the client will use the token provided in the `S2C_CONNECTION_LOGIN_RES ` for the next login attempt, when not quitting to desktop. And the newly generated token should also be in the database at this point.

This is as far as I can remember, Ideally I would like to keep this flow.
The logs should be a bit like this:
```
Received LoginToken:AAAA for platform:PC - first login
Updated OneTimeToken:BBBB - moved to game server
- return to title
Received LoginToken:BBBB for platform:PC -2nd login whiteout quitting to desktop
```

Would you be able to verify if the token that is set here: https://github.com/sebastian-heinz/Arrowgene.DragonsDogmaOnline/blob/develop/Arrowgene.Ddon.GameServer/Handler/ConnectionLoginHandler.cs#L125 is used during the next login request towards `ClientLoginHandler` or not?

Case 1)
If it is used then the current flow should be correct, and we need to understand why it is not working (perhaps the new key is not correctly sync in database?)

Case 2)
If the key set during that response is not used for the next login, but the old one, then it might be correct to comment out this line, in fact then we would not need the whole logic of generating a new key at all - but this begs the question what is the purpose of the response?

Case 3) the observation is inconsistent, sometimes it uses the old key, sometimes the new key set. in this case we need to understand the conditions when the client uses the old key vs the new key, so we only update the new key in appropriate case

Reply:
It's much more than I thought, no wonder why it took me so much time to figure out what was going on, anyway here's what I observated from my testing:

When the account is created no login token is generated.
After logging in through the launcher the login token is generated and tied to the proper account.
When hit the "Game Start" in the main title it will show in the log the recieved login token which is the previous generated from the above.
From the character selection screen if returned to main title then hit "Game Start" it will be able to connect to the same account with the previous login token.
This is the tricky part now, from the character selection screen (it doesn't matter if it's a new character), when the character is selected a SessionKey is generated and right after the OneTikeToken is updated, without the commented line it will overwrite the login token in the db.
When in the server selection screen if cancelled and returned to character screen it will either throw an error or make a new account depending on the setting. (With commented line the same account will still be accessed).
When ingame return to main title nothing will trigger, When hit "Game Start" a new login token is generated and since the last one was overwritten by the one above, it will not match the previous used to log in resulting in the error or creating a new account depending on the setting. (Since the SessionKey does not overwrite the OneTimeToken with the commented line, this issue doesn't happen and so the same account is accessed).
Lastly if logged in through the launcher the login token is regenerated and overwrites the old login token tied to the proper account and then repeat the loop.

These are my notes from what happens. I think that this is the first case but it shouldn't update the login token after selecting the character, I can answer why it's not the second case, everything is being called because to enter with selected character it requires the game token, so commenting the whole "update token" block the game will freeze in a black screen because it has no game token generated, so it's ideally functional with the commented line, though I don't know if that could cause any issues, but from what you stated it should be fine for now.

Reply:
thanks for your time to validate the behavior!